x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/SearchableSnapshotDirectoryTests.java

```diff
@@ -0,0 +1,361 @@
+import org.elasticsearch.cluster.metadata.RepositoryMetaData;
+import org.elasticsearch.common.CheckedBiConsumer;
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.common.lease.Releasables;
+import org.elasticsearch.common.lucene.BytesRefs;
+import org.elasticsearch.common.lucene.Lucene;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.ByteSizeUnit;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.core.internal.io.IOUtils;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.shard.ShardPath;
+import org.elasticsearch.index.snapshots.IndexShardSnapshotStatus;
+import org.elasticsearch.plugins.IndexStorePlugin.DirectoryFactory;
+import org.elasticsearch.repositories.IndexId;
+import org.elasticsearch.repositories.RepositoriesService;
+import org.elasticsearch.repositories.blobstore.BlobStoreRepository;
+import org.elasticsearch.repositories.blobstore.BlobStoreTestUtil;
+import org.elasticsearch.repositories.fs.FsRepository;
+import org.elasticsearch.snapshots.SnapshotId;
+import org.elasticsearch.test.DummyShardLock;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.test.IndexSettingsModule;
+import org.elasticsearch.threadpool.TestThreadPool;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots;
+
+import java.io.Closeable;
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasSize;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class SearchableSnapshotDirectoryTests extends ESTestCase {
+
+    public void testListAll() throws Exception {
+        testDirectories((directory, searchableDirectory) ->
+            assertThat(searchableDirectory.listAll(), equalTo(Arrays.stream(directory.listAll())
+                .filter(file -> "write.lock".equals(file) == false)
+                .filter(file -> file.startsWith("extra") == false)
+                .toArray(String[]::new))));
+    }
+
+    public void testFileLength() throws Exception {
+        testDirectories((directory, searchableDirectory) ->
+            Arrays.stream(directory.listAll())
+                .filter(file -> "write.lock".equals(file) == false)
+                .filter(file -> file.startsWith("extra") == false)
+                .forEach(file -> {
+                    try {
+                        assertThat(searchableDirectory.fileLength(file), equalTo(directory.fileLength(file)));
```

Comment:
Could we add a message including the filename to this assertion?

Reply:
Sure